### DISCUSSION TITLE
Error during the parsing with a readonly schema.

### DISCUSSION THREAD

**Discussion by vtgn** [colinhacks/zod#2795] [Q&A] [answered]
Hi,

I've been very disappointed by the fact that we can't parse an object with a readonly z.object schema: indeed, it is not secure to not freeze a public complex schema, used by several functionnalities of a software. A developer may do a mistake by modifying the internal data of the schema somewhere, which impacts directly the rest of the software functionnalities. This is not safe coding. :/

The error message during the execution of the Schema.parse method is :
`TypeError: Cannot assign to read only property '_cached' of object '#<O>'

      at O._getCached (../node_modules/@mylib.js:1:44278)
      at O._parse (../node_modules/@mylib.js:1:44518)
      at O._parseSync (../node_modules/@mylib.js:1:22947)
      at O.safeParse (../node_modules/@mylib.js:1:23438)
      at O.parse (../node_modules/@mylib.js:1:23121)`

I suppose the zod library stores a cache inside the z.object schema, which causes the problem. The library should store nothing inside the schema after its creation. If a cache must be used, why not store its contents somewhere inside an internal Map in the zod library inner files, associating a schema to the value(s) to cache associated to it? Plus the advantage you can manage a global cache, with a limited size and/or with a time-limited storage.

zod version: 3.21.4

Regards.

**Comment by vtgn** [comment]
Simply define a Zod schema, put it as argument in the `Object.freeze` method, define the Zod type of this schema, and parse an object with it:

```ts
// Create a zod schema and freeze it
const idZodSchema = Object.freeze( z.object( {
    id: z.number().positive().int().safe()
} ) )

// Create a zod type
type idZodType = z.infer<typeof idZodSchema>

// Parse an object with this zod type
const o: idZodType = idZodSchema.parse( { id: 3 } )
// ======> Error: Cannot assign to read only property '_cached' of object '#<ZodObject>'

console.log( o )
```

The complete error:
![image](https://github.com/colinhacks/zod/assets/112889052/6186e71f-9f6b-42a1-818d-804a4b44535f)

**Reply by JacobWeisenburger** [reply, marked as answer]
Zod is built with immutability in mind. So can you explain more about why you are trying to `Object.freeze` the schema?

###### https://github.com/colinhacks/zod#introduction
> Immutable: methods (e.g. `.optional()`) return a new instance

**Reply by vtgn** [reply]
I thought ZodObject was not immutable because of methods like setKey() but it seems it returns a new ZodObject. This naming misled me because the "set" verb means the inside object is modified, but it seems this is not the case here.

However, it is still possible to add properties to a schema or to overwrite the ones existing easily, causing possibly problems, but the risk is a little less important indeed.

It is just a little shame that we can't freeze the schemas to avoid some coding errors. I often freeze my objects to practice safe coding like real object languages.

Regards.
